File: 0695-max-area-of-island/0695-max-area-of-island.py
from collections import deque
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        
        rows = len(grid)
        cols = len(grid[0])
        directions = [(0, -1), (0, 1), (1, 0), (-1, 0)] #Up, Down, Right, Left
        visited = set()
        maxCount = 0

        def bfs(r, c, visited, directions):

            queue = deque()
            queue.append((r, c))
            visited.add((r, c))
            area = 1

            while queue:
                row, col = queue.popleft()

                #Check count?

                for nr, nc in directions:

                    newRow = nr + row
                    newCol = nc + col

                    #Check bounds
                    if 0 <= newRow < rows and 0 <= newCol < cols and grid[newRow][newCol] == 1 and (newRow, newCol) not in visited:
                        queue.append((newRow, newCol))
                        visited.add((newRow, newCol))
                        area += 1
            
            return area



        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == 1 and (row, col) not in visited:
                    x = bfs(row, col, visited, directions)
                    maxCount = max(maxCount, x)

        return maxCount

File: 0695-max-area-of-island/0695-max-area-of-island.cpp
#include <set>
#include <tuple>
#include <queue>

class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {

        int rows = grid.size();
        int cols = grid[0].size();

        //Here we loop over the array...
        for (int row = 0; row < rows; row++)
        {
            for (int col = 0; col < cols; col++)
            {

                if (grid[row][col] == 1 && hashy.find({row, col}) == hashy.end())
                {
                   std::cout << "xd" << std::endl; 
                   bfs(grid, row, col, cols, rows);
                   //count++;
                }
            }
        }

        return maxArea;
            
    }

    void bfs(vector<vector<int>>& grid, int row, int col, int cols, int rows)
    {
        int area = 0;
        queue.push({row, col});
        hashy.insert({row, col});

        while (!queue.empty())
        {
            //1. I want to add to hashy
            //2. Make adjuystments in bfs.. Which we will see further and add to queue.

            std::tuple<int, int> element = queue.front();
            queue.pop();

            //Split up element into row & col
            int nRow = std::get<0>(element);
            int nCol = std::get<1>(element);
            area++;
            //int nArea = std::get<2>(element);

            //Update our max
            cout << maxArea << std::endl;
            maxArea = std::max(maxArea, area);

            for (auto [dRow, dCol] : directions)
            {
                //Getting our newRow, newCol
                int newRow = nRow + dRow;
                int newCol = nCol + dCol;

                //Check our bounds && check if grid[row][col] == 1 && not in hashy.

                if ( 0 <= newRow && newRow < rows &&
                     0 <= newCol && newCol < cols &&
                     grid[newRow][newCol] == 1 &&
                     hashy.find({newRow, newCol}) == hashy.end())
                {
                    //1. Wanna add its neighbor to q
                    //2. Want to add to hashy to keep track of seen vals.
                    hashy.insert({newRow, newCol});
                    queue.push({newRow, newCol});

                }
            }
        }
    }

private:

    int maxArea = 0;
    int count = 0;
    std::set<std::tuple<int, int>> hashy; 
    std::queue<std::tuple<int, int>> queue;
    vector<tuple<int, int>> directions = {
        {0, -1},  // north
        {0, 1},   // south
        {1, 0},   // east
        {-1, 0}   // west
    };

};